<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no"><title>宏基因组分析 | 宋旭东</title><meta name="keywords" content="生物信息学,宏基因组"><meta name="author" content="Song Xudong"><meta name="copyright" content="Song Xudong"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#f7f9fe"><meta name="mobile-web-app-capable" content="yes"><meta name="apple-touch-fullscreen" content="yes"><meta name="apple-mobile-web-app-title" content="宏基因组分析"><meta name="application-name" content="宏基因组分析"><meta name="apple-mobile-web-app-capable" content="yes"><meta name="apple-mobile-web-app-status-bar-style" content="#f7f9fe"><meta property="og:type" content="article"><meta property="og:title" content="宏基因组分析"><meta property="og:url" content="http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/index.html"><meta property="og:site_name" content="宋旭东"><meta property="og:description" content="先使用rna-seq的环境 conda activate rna_p3        创建metagenomic分析环境 conda create -n metagenomicconda activate metagenomic  安装kneaddata conda install -c bioba"><meta property="og:locale" content="zh-CN"><meta property="og:image" content="http://example.com/picture/fengmian/hong.jpg"><meta property="article:author" content="Song Xudong"><meta property="article:tag"><meta name="twitter:card" content="summary"><meta name="twitter:image" content="http://example.com/picture/fengmian/hong.jpg"><meta name="description" content="先使用rna-seq的环境 conda activate rna_p3        创建metagenomic分析环境 conda create -n metagenomicconda activate metagenomic  安装kneaddata conda install -c bioba"><link rel="shortcut icon" href="/favicon.ico"><link rel="canonical" href="http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/"><link rel="preconnect" href="//cdn.cbd.int"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><meta name="google-site-verification" content="xxx"/><meta name="baidu-site-verification" content="code-xxx"/><meta name="msvalidate.01" content="xxx"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.cbd.int/node-snackbar@0.1.16/dist/snackbar.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.cbd.int/@fancyapps/ui@5.0.28/dist/fancybox/fancybox.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = {
  linkPageTop: undefined,
  peoplecanvas: undefined,
  postHeadAiDescription: undefined,
  diytitle: {"enable":true,"leaveTitle":"个人主页","backTitle":"个人主页"},
  LA51: undefined,
  greetingBox: undefined,
  twikooEnvId: '',
  commentBarrageConfig:undefined,
  music_page_default: "nav_music",
  root: '/',
  preloader: {"source":3},
  friends_vue_info: undefined,
  navMusic: false,
  mainTone: undefined,
  authorStatus: {"skills":["🧬 生物信息研究生","🎲 桌游爱好者","🦉 喜欢萌宠","🔍 分享与热心帮助"]},
  algolia: undefined,
  localSearch: undefined,
  translate: {"defaultEncoding":2,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简","rightMenuMsgToTraditionalChinese":"转为繁体","rightMenuMsgToSimplifiedChinese":"转为简体"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":330},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    simplehomepage: true,
    post: false
  },
  runtime: '天',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: {"copy":true,"copyrightEbable":false,"limitCount":50,"languages":{"author":"作者: Song Xudong","link":"链接: ","source":"来源: 宋旭东","info":"著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。","copySuccess":"复制成功，复制和转载请标注本文地址"}},
  lightbox: 'fancybox',
  Snackbar: {"chs_to_cht":"你已切换为繁体","cht_to_chs":"你已切换为简体","day_to_night":"你已切换为深色模式","night_to_day":"你已切换为浅色模式","bgLight":"#425AEF","bgDark":"#1f1f1f","position":"top-center"},
  source: {
    justifiedGallery: {
      js: 'https://cdn.cbd.int/flickr-justified-gallery@2.1.2/dist/fjGallery.min.js',
      css: 'https://cdn.cbd.int/flickr-justified-gallery@2.1.2/dist/fjGallery.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: true,
  isAnchor: false,
  shortcutKey: undefined,
  autoDarkmode: true
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  configTitle: '宋旭东',
  title: '宏基因组分析',
  postAI: '',
  pageFillDescription: '下载数据, 查看后台任务, SRA文件转为FASTQ格式, 单个转格式, 批量转格式, 质控, fastp质控, multiqc汇总质控结果, 去宿主-方法一, 构建索引, bowtie2比对单个, bowtie2比对批量, 使用hisat2直接用构建好的索引进行比对, 去宿主-方法二  kneaddata, 物种注释：kraken2, 单个比对, 批量比对, 物种组成及丰度估计, 结果整理, a多样性, 物种分类结果整理, 得到每个水平的结果, 基因组组装, 批量组装, Lefse, 在线绘制, R绘制先使用的环境创建分析环境安装下载数据参考这是一个人类的结直肠癌的宏基因组数据我们选择其中的个样进行分析需要数据的文件名使用下载数据该软件在的流程中有选择样本的可以加选择输出路径目录保存文件的目录直接在超算中输入命令下载并没有使用提交作业命令查看后台任务或者文件转为格式单个转格式慢的转换太慢了不建议使用将当前使用这个转换多线程转换格式输出的为的文件批量转格式小命令删除当前目录文件夹里的所有分文件夹只保留其文件进行批量转换将所有文件都放在文件夹里设置输入目录设置输出目录遍历目录中的所有文件获取不带路径的文件名使用处理每个文件生成在当前路径下的文件质控也只是类似的序列文件质控流程和转录组相似质控此处为扩展学习下载单个输入双端测序文件输出质控处理后文件和文件结果批量创建清理后的文件夹设置工作目录为文件所在的目录遍历所有以结尾的文件从文件名中提取没有的部分构建对应的文件名在后台执行检查和命令检查对应的文件是否存在如果存在执行命令如果不存在打印错误信息等待所有后台进程完成汇总质控结果这个结果看不看无所谓是强大的质控软件只要输入文件无误结果也不会有问题结尾的文件即为之后进行分析的清洗之后的序列文件去宿主方法一去宿主的过程其实就是将序列比对到宿主基因组上然后没有比对到的序列整合成新文件就是去宿主后的了宿主基因组需要自己先下载好并用建立索引以人类为例构建索引在官网中找到自己的物种人的基因组鼠的基因组自己构建索引有点慢比对单个参考比对批量将工作目录设置为文件所在的目录将传入的参数赋值给变量遍历所有以第一个参数模式结尾的文件从文件名中提取去掉模式后的部分构建对应的第二个参数模式的文件名在后台执行检查和命令检查对应的文件是否存在找到名为的文件对应如果存在例如执行命令如果不存在打印错误信息错误未找到与匹配的文件等待所有后台进程完成使用直接用构建好的索引进行比对尝试使用建立好的索引进行比对参考基因组的位置假设你的文件在文件夹中遍历文件夹中的所有文件从文件名中提取并删除查找对应的文件检查文件是否存在使用进行比对并指定输出目录为当前目录指定输出目录为当前目录去宿主方法二使用进行质控和去宿主自带的参考序列有限主要为人和小鼠其他的需要自己构建查看有哪些数据库下载人的参考序列速度似乎还可以物种注释参考是一个用于对高通量测序数据进行分类和标识物种的软件它使用参考数据库中的基因组序列来进行分类并使用方法来实现快速和准确的分类使用进行基本分类的简单步骤安装可以从官方网站下载并安装软件准备参考数据库需要一个参考数据库以便对测序数据进行分类可以直接下载官方构建的标准库也可以从或其他数据库下载相应的基因组序列并使用内置的工具来构建数据库标准模式下只下载种数据库古菌细菌人类载体病毒超算中未成功选择自行下载选择自行下载网站下载文件大小自行下载命令运行使用对测序数据进行分类需要使用以下命令这里是参考数据库的路径是需要进行分类的输入文件是输出文件的名称将输出一个分类报告文件和一个序列文件需要注意的是运行至少要提供数据库大小的内存大小运行内存因为它会把整个数据库载入内存后进行序列的注释所以如果发现无法载入数据库的报错可以尝试调大内存资源单个比对第二行为存放参考索引的文件夹批量比对将工作目录设置为文件所在的目录将传入的参数赋值给变量遍历所有以第一个参数模式结尾的文件从文件名中提取去掉模式后的部分构建对应的第二个参数模式的文件名在后台执行检查和命令检查对应的文件是否存在找到名为的文件对应如果存在例如执行命令文件运行完毕如果不存在打印错误信息错误未找到与匹配的文件等待所有后台进程完成物种组成及丰度估计只能说宏基因组的教程也太水了未找到对结果的处理部分使用教程运行进行各个分类水平物种丰度估计先安装运行进行各个分类水平物种丰度估计将工作目录设置为文件所在的目录将传入的参数赋值给变量遍历所有以第一个参数模式结尾的文件从文件名中提取去掉模式后的部分直接执行命令不需要检查对应的是否存在找到名为的文件循环执行代码区运行等待所有后台进程完成结果整理安装文件合并成格式转表格注意这里假设是指向转换为的工具不是的如果是的那么下面的命令是错误的正确的命令应该是直接使用工具进行转换输出文件格式调整补全物种名替换回拉丁名保留丰度信息用于后续绘图输出结果为包含物种分类信息的丰度文件可以用此结果绘制物种堆积图为去除物种分类信息的丰度文件暂未准备绘图代码多样性教程中未具体讲解方法但是提到使用及两个包进行多样性分析使用做过的方法进行分析文件分组文件语言安装和加载必要的包抽平求和查看每个样本的和使用该代码进行抽平查看抽平后的每个样本的和物种分类结果整理还是使用之前的方法使用上面的为包含物种分类信息的丰度文件可以用此结果绘制物种堆积图保存导出分类好的文件得到每个水平的结果界门纲目科属种只能注释到属水平的结果虽然我们有属水平的文件但是是没有相应的可用结果载入必要的库读取数据移除列删除等于的行分类级别列表上一层级名称遍历每个分类级别进行分组和求和并保存到文件确保当前层级和上一层级列存在并且数据类型正确对当前层级列进行分组并求和同时保留上一层级信息生成文件名保存结果到文件更新上一层级为当前层级基因组组装使用进行组装安装单个的示例输入输入最小长度设置目录内存占用线程数输出目录输出前缀多组数据组装输入数据逗号分隔批量组装要求输出文件夹会自动创建开始时不需要存在将工作目录设置为文件所在的目录将传入的参数赋值给变量遍历所有以第一个参数模式结尾的文件从文件名中提取去掉模式后的部分构建对应的第二个参数模式的文件名在后台执行检查和命令检查对应的文件是否存在找到名为的文件对应如果存在例如执行命令如果不存在打印错误信息错误未找到与匹配的文件等待所有后台进程完成参考需要个输入文件过少可增大下面选项保存保存分析出初步结果可自行画图输入文件可在在线网站进行可视化或者可视化在线绘制复制内容进行修改找到选项主要是更改第一行的列名定义分组检查数据可用后可直接出结果某些未知的菌可在原始数据中提前删除在线网站的编辑功能好像更强大了作图方便建议采用绘制使用包自带的绘图并不算好看可进行美化或使用原始数据自行绘图为显示的个数为自己的分组',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-11-22 14:05:24',
  postMainColor: '',
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
      win.saveToLocal = {
        set: (key, value, ttl) => {
          if (ttl === 0) return
          const now = Date.now()
          const expiry = now + ttl * 86400000
          const item = {
            value,
            expiry
          }
          localStorage.setItem(key, JSON.stringify(item))
        },
      
        get: key => {
          const itemStr = localStorage.getItem(key)
      
          if (!itemStr) {
            return undefined
          }
          const item = JSON.parse(itemStr)
          const now = Date.now()
      
          if (now > item.expiry) {
            localStorage.removeItem(key)
            return undefined
          }
          return item.value
        }
      }
    
      win.getScript = (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        script.onerror = reject
        script.onload = script.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          script.onload = script.onreadystatechange = null
          resolve()
        }

        Object.keys(attr).forEach(key => {
          script.setAttribute(key, attr[key])
        })

        document.head.appendChild(script)
      })
    
      win.getCSS = (url, id = false) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onerror = reject
        link.onload = link.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          link.onload = link.onreadystatechange = null
          resolve()
        }
        document.head.appendChild(link)
      })
    
      win.activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#18171d')
        }
      }
      win.activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#f7f9fe')
        }
      }
      const t = saveToLocal.get('theme')
    
          const isDarkMode = window.matchMedia('(prefers-color-scheme: dark)').matches
          const isLightMode = window.matchMedia('(prefers-color-scheme: light)').matches
          const isNotSpecified = window.matchMedia('(prefers-color-scheme: no-preference)').matches
          const hasNoSupport = !isDarkMode && !isLightMode && !isNotSpecified

          if (t === undefined) {
            if (isLightMode) activateLightMode()
            else if (isDarkMode) activateDarkMode()
            else if (isNotSpecified || hasNoSupport) {
              const now = new Date()
              const hour = now.getHours()
              const isNight = hour <= 6 || hour >= 18
              isNight ? activateDarkMode() : activateLightMode()
            }
            window.matchMedia('(prefers-color-scheme: dark)').addListener(e => {
              if (saveToLocal.get('theme') === undefined) {
                e.matches ? activateDarkMode() : activateLightMode()
              }
            })
          } else if (t === 'light') activateLightMode()
          else activateDarkMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
      const detectApple = () => {
        if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
          document.documentElement.classList.add('apple')
        }
      }
      detectApple()
    })(window)</script><meta name="generator" content="Hexo 7.3.0"></head><body data-type="anzhiyu"><div id="web_bg"></div><div id="an_music_bg"></div><div id="loading-box" onclick="document.getElementById(&quot;loading-box&quot;).classList.add(&quot;loaded&quot;)"><div class="loading-bg"><img class="loading-img nolazyload" alt="加载头像" src="/img/touxiang.png"/><div class="loading-image-dot"></div></div></div><script>const preloader = {
  endLoading: () => {
    document.getElementById('loading-box').classList.add("loaded");
  },
  initLoading: () => {
    document.getElementById('loading-box').classList.remove("loaded")
  }
}
window.addEventListener('load',()=> { preloader.endLoading() })
setTimeout(function(){preloader.endLoading();},10000)

if (true) {
  document.addEventListener('pjax:send', () => { preloader.initLoading() })
  document.addEventListener('pjax:complete', () => { preloader.endLoading() })
}</script><link rel="stylesheet" href="https://cdn.cbd.int/anzhiyu-theme-static@1.1.10/progress_bar/progress_bar.css"/><script async="async" src="https://cdn.cbd.int/pace-js@1.2.4/pace.min.js" data-pace-options="{ &quot;restartOnRequestAfter&quot;:false,&quot;eventLag&quot;:false}"></script><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><div id="nav-group"><span id="blog_name"><a id="site-name" href="/" accesskey="h"><div class="title">宋旭东</div><i class="anzhiyufont anzhiyu-icon-house-chimney"></i></a></span><div class="mask-name-container"><div id="name-container"><a id="page-name" href="javascript:anzhiyu.scrollToDest(0, 500)">PAGE_NAME</a></div></div><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> Blog</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/archives/"><i class="anzhiyufont anzhiyu-icon-box-archive faa-tada" style="font-size: 0.9em;"></i><span> 总览</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/categories/"><i class="anzhiyufont anzhiyu-icon-shapes faa-tada" style="font-size: 0.9em;"></i><span> 分类</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/tags/"><i class="anzhiyufont anzhiyu-icon-tags faa-tada" style="font-size: 0.9em;"></i><span> 标签</span></a></li></ul></div><div class="menus_item"><a class="site-page faa-parent animated-hover" href="/about/"><span> 关于本人</span></a></div><div class="menus_item"><a class="site-page faa-parent animated-hover" href="/comments/"><span> 留言板</span></a></div></div></div><div id="nav-right"><div class="nav-button" id="randomPost_button"><a class="site-page" onclick="toRandomPost()" title="随机前往一个文章" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-dice"></i></a></div><input id="center-console" type="checkbox"/><label class="widget" for="center-console" title="中控台" onclick="anzhiyu.switchConsole();"><i class="left"></i><i class="widget center"></i><i class="widget right"></i></label><div id="console"><div class="console-card-group-reward"><ul class="reward-all console-card"><li class="reward-item"><a href="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-weichat.png" target="_blank"><img class="post-qr-code-img" alt="微信" src="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-weichat.png"/></a><div class="post-qr-code-desc">微信</div></li><li class="reward-item"><a href="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-alipay.png" target="_blank"><img class="post-qr-code-img" alt="支付宝" src="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-alipay.png"/></a><div class="post-qr-code-desc">支付宝</div></li></ul></div><div class="console-card-group"><div class="console-card-group-left"><div class="console-card" id="card-newest-comments"><div class="card-content"><div class="author-content-item-tips">互动</div><span class="author-content-item-title"> 最新评论</span></div><div class="aside-list"><span>正在加载中...</span></div></div></div><div class="console-card-group-right"><div class="console-card tags"><div class="card-content"><div class="author-content-item-tips">兴趣点</div><span class="author-content-item-title">寻找你感兴趣的领域</span><div class="card-tags"><div class="item-headline"></div><div class="card-tag-cloud"><a href="/tags/AI/" style="font-size: 1.05rem;">AI<sup>1</sup></a><a href="/tags/MATLAB/" style="font-size: 1.05rem;">MATLAB<sup>3</sup></a><a href="/tags/RNA-seq/" style="font-size: 1.05rem;">RNA-seq<sup>1</sup></a><a href="/tags/Ribo-seq/" style="font-size: 1.05rem;">Ribo-seq<sup>1</sup></a><a href="/tags/%E4%B8%80%E4%BD%9C/" style="font-size: 1.05rem;">一作<sup>1</sup></a><a href="/tags/%E4%BA%8C%E4%BD%9C/" style="font-size: 1.05rem;">二作<sup>1</sup></a><a href="/tags/%E5%88%86%E5%AD%90%E5%AF%B9%E6%8E%A5/" style="font-size: 1.05rem;">分子对接<sup>1</sup></a><a href="/tags/%E5%A4%A7%E5%AD%A6/" style="font-size: 1.05rem;">大学<sup>1</sup></a><a href="/tags/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 1.05rem;">宏基因组<sup>1</sup></a><a href="/tags/%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86/" style="font-size: 1.05rem;">影像数据处理<sup>3</sup></a><a href="/tags/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0/" style="font-size: 1.05rem;">机器学习<sup>1</sup></a><a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" style="font-size: 1.05rem;">生物信息学<sup>4</sup></a><a href="/tags/%E8%8E%B7%E5%A5%96/" style="font-size: 1.05rem;">获奖<sup>1</sup></a><a href="/tags/%E8%AE%BA%E6%96%87/" style="font-size: 1.05rem;">论文<sup>2</sup></a></div></div><hr/></div></div><div class="console-card history"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-box-archiv"></i><span>文章</span></div><div class="card-archives"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-archive"></i><span>归档</span></div><ul class="card-archive-list"><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2025/03/"><span class="card-archive-list-date">三月 2025</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">2</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2025/02/"><span class="card-archive-list-date">二月 2025</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">1</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2025/01/"><span class="card-archive-list-date">一月 2025</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">2</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/11/"><span class="card-archive-list-date">十一月 2024</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">6</span><span>篇</span></div></a></li></ul></div><hr/></div></div></div><div class="button-group"><div class="console-btn-item"><a class="darkmode_switchbutton" title="显示模式切换" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-moon"></i></a></div><div class="console-btn-item" id="consoleHideAside" onclick="anzhiyu.hideAsideBtn()" title="边栏显示控制"><a class="asideSwitch"><i class="anzhiyufont anzhiyu-icon-arrows-left-right"></i></a></div></div><div class="console-mask" onclick="anzhiyu.hideConsole()" href="javascript:void(0);"></div></div><div class="nav-button" id="nav-totop"><a class="totopbtn" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i><span id="percent" onclick="anzhiyu.scrollToDest(0,500)">0</span></a></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);" title="切换"><i class="anzhiyufont anzhiyu-icon-bars"></i></a></div></div></div></nav><div id="post-info"><div id="post-firstinfo"><div class="meta-firstline"><a class="post-meta-original">原创</a><span class="post-meta-categories"><span class="post-meta-separator"></span><i class="anzhiyufont anzhiyu-icon-inbox post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" itemprop="url">生物信息学</a></span><span class="article-meta tags"><a class="article-meta__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" tabindex="-1" itemprop="url"> <span> <i class="anzhiyufont anzhiyu-icon-hashtag"></i>生物信息学</span></a><a class="article-meta__tags" href="/tags/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/" tabindex="-1" itemprop="url"> <span> <i class="anzhiyufont anzhiyu-icon-hashtag"></i>宏基因组</span></a></span></div></div><h1 class="post-title" itemprop="name headline">宏基因组分析</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="anzhiyufont anzhiyu-icon-calendar-days post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" itemprop="dateCreated datePublished" datetime="2024-11-22T05:54:04.889Z" title="发表于 2024-11-22 13:54:04">2024-11-22</time><span class="post-meta-separator"></span><i class="anzhiyufont anzhiyu-icon-history post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" itemprop="dateCreated datePublished" datetime="2024-11-22T06:05:24.164Z" title="更新于 2024-11-22 14:05:24">2024-11-22</time></span></div><div class="meta-secondline"><span class="post-meta-separator"></span><span class="post-meta-wordcount"><i class="anzhiyufont anzhiyu-icon-file-word post-meta-icon" title="文章字数"></i><span class="post-meta-label" title="文章字数">字数总计:</span><span class="word-count" title="文章字数">4.6k</span><span class="post-meta-separator"></span><i class="anzhiyufont anzhiyu-icon-clock post-meta-icon" title="阅读时长"></i><span class="post-meta-label" title="阅读时长">阅读时长:</span><span>20分钟</span></span><span class="post-meta-separator">       </span><span class="post-meta-position" title="作者IP属地为南宁"><i class="anzhiyufont anzhiyu-icon-location-dot"></i>南宁</span></div></div></div><section class="main-hero-waves-area waves-area"><svg class="waves-svg" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" viewBox="0 24 150 28" preserveAspectRatio="none" shape-rendering="auto"><defs><path id="gentle-wave" d="M -160 44 c 30 0 58 -18 88 -18 s 58 18 88 18 s 58 -18 88 -18 s 58 18 88 18 v 44 h -352 Z"></path></defs><g class="parallax"><use href="#gentle-wave" x="48" y="0"></use><use href="#gentle-wave" x="48" y="3"></use><use href="#gentle-wave" x="48" y="5"></use><use href="#gentle-wave" x="48" y="7"></use></g></svg></section><div id="post-top-cover"><img class="nolazyload" id="post-top-bg" src="/picture/fengmian/hong.jpg"></div></header><main id="blog-container"><div class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container" itemscope itemtype="http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/"><header><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" itemprop="url">生物信息学</a><a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" tabindex="-1" itemprop="url">生物信息学</a><a href="/tags/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/" tabindex="-1" itemprop="url">宏基因组</a><h1 id="CrawlerTitle" itemprop="name headline">宏基因组分析</h1><span itemprop="author" itemscope itemtype="http://schema.org/Person">Song Xudong</span><time itemprop="dateCreated datePublished" datetime="2024-11-22T05:54:04.889Z" title="发表于 2024-11-22 13:54:04">2024-11-22</time><time itemprop="dateCreated datePublished" datetime="2024-11-22T06:05:24.164Z" title="更新于 2024-11-22 14:05:24">2024-11-22</time></header><p>先使用rna-seq的环境</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda activate rna_p3      </span><br></pre></td></tr></table></figure>

<p>创建metagenomic分析环境</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda create -n metagenomic</span><br><span class="line"></span><br><span class="line">conda activate metagenomic</span><br></pre></td></tr></table></figure>

<p>安装kneaddata</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda install -c biobakery kneaddata</span><br></pre></td></tr></table></figure>



<h1 id="下载数据"><a href="#下载数据" class="headerlink" title="下载数据"></a>下载数据</h1><p>参考： <a target="_blank" rel="noopener" href="https://blog.csdn.net/Mr_pork/article/details/139743229">https://blog.csdn.net/Mr_pork/article/details/139743229</a></p>
<p>这是一个  人类的  结直肠癌 的宏基因组数据，我们选择其中的10个样进行分析</p>
<p>需要数据的文件名，使用prefetch 下载数据，该软件在rna-seq的流程中有</p>
<p>SRA.txt</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240831151004127.png" alt="image-20240831151004127"></p>
<p>选择样本的metadata</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240831151030690.png" alt="image-20240831151030690"></p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">nohup prefetch -f no --option-file SRA.txt &amp;</span><br></pre></td></tr></table></figure>

<p>可以加 -O 选择输出路径</p>
<ul>
<li><code>-O|--output-directory &lt;目录&gt;</code>：保存文件的目录。</li>
</ul>
<p>直接在超算中输入命令下载，并没有使用sbatch提交作业命令</p>
<h3 id="查看后台任务"><a href="#查看后台任务" class="headerlink" title="查看后台任务"></a>查看后台任务</h3><figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">jobs</span><br><span class="line"></span><br><span class="line">#或者</span><br><span class="line">ps -f</span><br></pre></td></tr></table></figure>

<h1 id="SRA文件转为FASTQ格式"><a href="#SRA文件转为FASTQ格式" class="headerlink" title="SRA文件转为FASTQ格式"></a>SRA文件转为FASTQ格式</h1><h2 id="单个转格式"><a href="#单个转格式" class="headerlink" title="单个转格式"></a>单个转格式</h2><p>慢的转换，太慢了，不建议使用</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">#将当前</span><br><span class="line">fastq-dump --split-3 --gzip ./SRR12207279</span><br></pre></td></tr></table></figure>

<p>使用这个转换，多线程转换格式，输出的为fq的文件</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">fasterq-dump --split-3 ./SRR12207283</span><br></pre></td></tr></table></figure>

<h2 id="批量转格式"><a href="#批量转格式" class="headerlink" title="批量转格式"></a>批量转格式</h2><p>小命令：删除当前目录SRR文件夹里的所有分文件夹，只保留其文件</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">find ./SRR -mindepth 1 -type d -exec sh -c &#x27;mv &#123;&#125;/* ./SRR; rmdir &#123;&#125;&#x27; \;</span><br></pre></td></tr></table></figure>

<p>fasterq-dump进行批量转换，将所有 .sra 文件都放在SRR文件夹里</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment"># 设置输入目录</span></span><br><span class="line">sra_dir<span class="operator">=</span><span class="string">&quot;./SRR&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置输出目录</span></span><br><span class="line">output_dir<span class="operator">=</span><span class="string">&quot;./fastq-result&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 遍历目录中的所有.sra文件</span></span><br><span class="line"><span class="keyword">for</span> sra_file <span class="keyword">in</span> <span class="operator">$</span>sra_dir<span class="operator">/</span><span class="operator">*</span>.sra</span><br><span class="line">do</span><br><span class="line">    <span class="comment"># 获取不带路径的文件名</span></span><br><span class="line">    filename<span class="operator">=</span><span class="operator">$</span><span class="punctuation">(</span>basename <span class="string">&quot;$sra_file&quot;</span> .sra<span class="punctuation">)</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 使用fasterq-dump处理每个文件</span></span><br><span class="line">    fasterq<span class="operator">-</span>dump <span class="operator">-</span><span class="operator">-</span>outdir <span class="string">&quot;$output_dir&quot;</span> <span class="operator">-</span><span class="operator">-</span>split<span class="operator">-</span><span class="number">3</span> <span class="string">&quot;$sra_file&quot;</span></span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p>生成在当前路径 .&#x2F;fastq-result 下的 fastq 文件</p>
<h1 id="质控"><a href="#质控" class="headerlink" title="质控"></a>质控</h1><p>也只是类似的序列文件，质控流程和转录组相似</p>
<h2 id="fastp质控"><a href="#fastp质控" class="headerlink" title="fastp质控"></a>fastp质控</h2><p>此处为扩展学习</p>
<p>conda下载fastp</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line"># note: the fastp version in bioconda may be not the latest</span><br><span class="line">conda install -c bioconda fastp</span><br></pre></td></tr></table></figure>

<p>单个</p>
<p>输入 -i -I 双端测序文件 ，输出 -o -O 质控处理后文件，和 json文件，fastp.html结果</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">fastp -i in.R1.fq.gz -I in.R2.fq.gz -o out.R1.fq.gz -O out.R2.fq.gz</span><br></pre></td></tr></table></figure>

<p>批量</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line"># 创建清理后的文件夹</span><br><span class="line">mkdir  clean-fastp</span><br><span class="line"></span><br><span class="line"># 设置工作目录为fastq文件所在的目录</span><br><span class="line">cd ./fastq-result/</span><br><span class="line"></span><br><span class="line"># 遍历所有以_1.fastq结尾的文件</span><br><span class="line">for file1 in *_1.fastq; do</span><br><span class="line">    # 从文件名中提取没有_1的部分</span><br><span class="line">    base=$(basename &quot;$file1&quot; _1.fastq)</span><br><span class="line">    </span><br><span class="line">    # 构建对应的_2.fastq文件名</span><br><span class="line">    file2=&quot;$&#123;base&#125;_2.fastq&quot;</span><br><span class="line">fileoo1=&quot;$&#123;base&#125;_1.fq&quot;</span><br><span class="line">fileoo2=&quot;$&#123;base&#125;_2.fq&quot;</span><br><span class="line">jsono=&quot;$&#123;base&#125;.json&quot;</span><br><span class="line">htmlo=&quot;$&#123;base&#125;.html&quot;</span><br><span class="line"></span><br><span class="line">    </span><br><span class="line">    # 在后台执行检查和trim_galore命令</span><br><span class="line">    (</span><br><span class="line">        # 检查对应的_2.fastq文件是否存在</span><br><span class="line">        if [ -e &quot;$file2&quot; ]; then</span><br><span class="line">            # 如果存在，执行trim_galore命令</span><br><span class="line">		fastp -i &quot;$file1&quot;  -o ../clean-fastp/&quot;$fileoo1&quot; -I &quot;$file2&quot; -O ../clean-fastp/&quot;$fileoo2&quot;  --json  ../clean-fastp/&quot;$jsono&quot;  --html  ../clean-fastp/&quot;$htmlo&quot;</span><br><span class="line">                    else</span><br><span class="line">            # 如果不存在，打印错误信息</span><br><span class="line">            echo &quot;Error: No matching file found for $file1&quot;</span><br><span class="line">        fi</span><br><span class="line">    ) &amp;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 等待所有后台进程完成</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>

<h2 id="multiqc汇总质控结果"><a href="#multiqc汇总质控结果" class="headerlink" title="multiqc汇总质控结果"></a>multiqc汇总质控结果</h2><figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">multiqc ./fastq-result/ -o   ./fastq-result/</span><br></pre></td></tr></table></figure>

<p>这个结果看不看无所谓，fastp是强大的质控软件，只要输入文件无误，结果也不会有问题</p>
<p>.fq  结尾的文件即为之后进行分析的，清洗之后的序列文件</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240831181023502.png" alt="image-20240831181023502"></p>
<h1 id="去宿主-方法一"><a href="#去宿主-方法一" class="headerlink" title="去宿主-方法一"></a>去宿主-方法一</h1><p>去宿主的过程其实就是将序列比对到宿主基因组上，然后没有比对到的序列整合成新文件就是去宿主后的了。宿主基因组需要自己先下载好并用 bowtie2-build 建立索引，以人类为例：</p>
<h2 id="构建索引"><a href="#构建索引" class="headerlink" title="构建索引"></a>构建索引</h2><p>在官网中找到自己的物种  <a target="_blank" rel="noopener" href="https://hgdownload.soe.ucsc.edu/downloads.html">https://hgdownload.soe.ucsc.edu/downloads.html</a></p>
<p>人的基因组 hg38</p>
<p><a target="_blank" rel="noopener" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chromFa.tar.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chromFa.tar.gz</a></p>
<p>鼠的基因组 mm10</p>
<p><a target="_blank" rel="noopener" href="http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz">http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz</a> </p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chromFa.tar.gz</span><br><span class="line">tar -zxvf chromFa.tar.gz </span><br><span class="line">cat *.fa &gt; hg38.fa</span><br><span class="line">bowtie2-build hg38.fa hg38</span><br></pre></td></tr></table></figure>

<p>自己构建索引有点慢</p>
<h2 id="bowtie2比对-单个"><a href="#bowtie2比对-单个" class="headerlink" title="bowtie2比对,单个"></a>bowtie2比对,单个</h2><p>参考  <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/fe9c5cc7373e">https://www.jianshu.com/p/fe9c5cc7373e</a></p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">bowtie2 -p 20 -x /public/home/dk_szy/songxudong/metagenomic/db/hg37dec_v0.1 -1 data/fastp/$&#123;sample&#125;_1.fq.gz \</span><br><span class="line">    -2 data/fastp/$&#123;sample&#125;_2.fq.gz -S data/rm_human/$&#123;sample&#125;.sam \</span><br><span class="line">    --un-conc data/rm_human/$&#123;sample&#125;.fq --very-sensitive</span><br><span class="line">  rm data/rm_human/$&#123;sample&#125;.sam</span><br></pre></td></tr></table></figure>

<h2 id="bowtie2比对-批量"><a href="#bowtie2比对-批量" class="headerlink" title="bowtie2比对,批量"></a>bowtie2比对,批量</h2><figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line"># 将工作目录设置为fastq文件所在的目录！！！！</span><br><span class="line">mkdir rm_human</span><br><span class="line"></span><br><span class="line">cd ./clean-fastp/</span><br><span class="line"></span><br><span class="line"># 将传入的参数赋值给变量！！！！！</span><br><span class="line">file1_pattern=&quot;_1.fq&quot;</span><br><span class="line">file2_pattern=&quot;_2.fq&quot;</span><br><span class="line"></span><br><span class="line"># 遍历所有以第一个参数模式结尾的文件</span><br><span class="line">for file1 in *$&#123;file1_pattern&#125;; do</span><br><span class="line">    # 从文件名中提取去掉模式后的部分</span><br><span class="line">    base=$(basename &quot;$file1&quot; $&#123;file1_pattern&#125;)</span><br><span class="line">    # 构建对应的第二个参数模式的文件名</span><br><span class="line">    file2=&quot;$&#123;base&#125;$&#123;file2_pattern&#125;&quot;</span><br><span class="line"></span><br><span class="line">    # 在后台执行检查和trim_galore命令</span><br><span class="line">    (</span><br><span class="line">        # 检查对应的文件是否存在</span><br><span class="line">        if [ -e &quot;$file2&quot; ]; then</span><br><span class="line">		echo &quot;找到名为 &quot;$base&quot; 的文件 $file1 对应 $file2 &quot;</span><br><span class="line">            # 如果存在，例如，执行trim_galore命令！！！！！！</span><br><span class="line">            bowtie2 -p 20 -x /public/home/dk_szy/songxudong/metagenomic/db/hg37dec_v0.1 -1 &quot;$file1&quot; \</span><br><span class="line">    -2  &quot;$file2&quot; -S $&#123;base&#125;.sam \</span><br><span class="line">    --un-conc ../rm_human/$&#123;base&#125;.fq --very-sensitive</span><br><span class="line">  rm $&#123;base&#125;.sam</span><br><span class="line">            </span><br><span class="line">            </span><br><span class="line">        else</span><br><span class="line">            # 如果不存在，打印错误信息</span><br><span class="line">            echo &quot;错误: 未找到与 $file1 匹配的文件&quot;</span><br><span class="line">        fi</span><br><span class="line">    ) &amp;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 等待所有后台进程完成</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>



<h2 id="使用hisat2直接用构建好的索引进行比对"><a href="#使用hisat2直接用构建好的索引进行比对" class="headerlink" title="#使用hisat2直接用构建好的索引进行比对"></a>#使用hisat2直接用构建好的索引进行比对</h2><p>尝试使用建立好的索引进行比对</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line"></span><br><span class="line">mkdir -p ./align/flag</span><br><span class="line">cd ./align/</span><br><span class="line">pwd</span><br><span class="line"></span><br><span class="line">##参考基因组的位置</span><br><span class="line">index=&#x27;/public/home/dk_szy/songxudong/rna-test/reference/human-UCSC-hg38/hg38/genome&#x27;</span><br><span class="line"></span><br><span class="line"># 假设你的fastq文件在fastq-result文件夹中</span><br><span class="line">fastq_dir=&quot;../clean-fastp&quot;</span><br><span class="line"></span><br><span class="line"># 遍历fastq-result文件夹中的所有1.fastq文件</span><br><span class="line">for file1 in $fastq_dir/*_1.fq; do</span><br><span class="line">    # 从1.fastq文件名中提取ID，并删除_1_val_1</span><br><span class="line">    id=$(basename &quot;$file1&quot; .fq | sed &#x27;s/_1_val_1//&#x27;)</span><br><span class="line">    </span><br><span class="line">    # 查找对应的2.fastq文件</span><br><span class="line">    file2=&quot;$fastq_dir/$&#123;id&#125;_2.fq&quot;</span><br><span class="line">    </span><br><span class="line">    # 检查2.fastq文件是否存在</span><br><span class="line">    if [ -f &quot;$file2&quot; ]; then</span><br><span class="line">        echo &quot;333#  $&#123;id&#125;  ！！！！！ is on the hisat2 Working !!!&quot;</span><br><span class="line">        </span><br><span class="line">        # 使用hisat2进行比对，并指定输出目录为当前目录（./align/）</span><br><span class="line">        hisat2 -t -p 20  -x $index \</span><br><span class="line">            -1 &quot;$file1&quot; \</span><br><span class="line">            -2 &quot;$file2&quot;  -S  &quot;$&#123;id&#125;.sam&quot; </span><br><span class="line">        </span><br><span class="line">        # sam2bam and remove sam，指定输出目录为当前目录（./align/）</span><br><span class="line">        echo -e &quot; $&#123;id&#125; sam2bam and remove sam   &quot;</span><br><span class="line">        samtools view -F 12 -@ 12 -b &quot;./$&#123;id&#125;.sam&quot; &gt; &quot;./$&#123;id&#125;_sorted.bam&quot;</span><br><span class="line">        rm &quot;./$&#123;id&#125;.sam&quot;</span><br><span class="line">    else</span><br><span class="line">        echo &quot;No matching 2.fastq file found for $file1&quot;</span><br><span class="line">    fi</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<h1 id="去宿主-方法二-kneaddata"><a href="#去宿主-方法二-kneaddata" class="headerlink" title="#去宿主-方法二  kneaddata"></a>#去宿主-方法二  kneaddata</h1><p>使用 kneaddata进行质控和去宿主</p>
<p>kneaddata自带的参考序列有限，主要为人和小鼠，其他的需要自己构建</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">#查看有哪些数据库</span><br><span class="line">kneaddata_database</span><br></pre></td></tr></table></figure>

<p>下载人的参考序列，速度似乎还可以</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">mkdir -p db</span><br><span class="line"></span><br><span class="line">kneaddata_database --download human_genome bowtie2 db/</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">kneaddata -i seq/C2_1.fq.gz -i seq/C2_2.fq.gz</span><br><span class="line"></span><br><span class="line"> -o qc/ -v -t 8 --remove-intermediate-output</span><br><span class="line"></span><br><span class="line"> --trimmomatic ~/.conda/envs/qc2/share/trimmomatic</span><br><span class="line"></span><br><span class="line"> --trimmomatic-options &#x27;ILLUMINACLIP:~/.conda/envs/qc2/share/trimmomatic/adapters/TruSeq3-PE.fa:2:40:15 SLIDINGWINDOW:4:20 MINLEN:50&#x27;</span><br><span class="line"></span><br><span class="line"> --bowtie2-options &#x27;--very-sensitive --dovetail&#x27;</span><br><span class="line"></span><br><span class="line"> --bowtie2-options=&quot;--reorder&quot;</span><br><span class="line"></span><br><span class="line"> -db db/Homo_sapiens</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h1 id="物种注释：kraken2"><a href="#物种注释：kraken2" class="headerlink" title="物种注释：kraken2"></a>物种注释：kraken2</h1><p>参考：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/fe9c5cc7373e">https://www.jianshu.com/p/fe9c5cc7373e</a></p>
<p>Kraken2是一个用于对高通量测序数据进行分类和标识物种的软件。它使用参考数据库中的基因组序列来进行分类，并使用k-mer方法来实现快速和准确的分类。</p>
<p>使用Kraken2进行基本分类的简单步骤：</p>
<p>安装Kraken2：可以从Kraken2官方网站下载并安装Kraken2软件。</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda install bioconda::kraken2</span><br></pre></td></tr></table></figure>

<p>准备参考数据库：Kraken2需要一个参考数据库，以便对测序数据进行分类。可以直接下载官方构建的标准库，也可以从NCBI、Ensembl或其他数据库下载相应的基因组序列，并使用Kraken2内置的工具来构建数据库。</p>
<p><code>--standard</code>标准模式下只下载5种数据库：古菌archaea、细菌bacteria、人类human、载体UniVec_Core、病毒viral。</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">#超算中未成功，选择自行下载</span><br><span class="line">#kraken2-build --standard --threads 20 --db ./</span><br></pre></td></tr></table></figure>

<p>选择自行下载网站：  <a target="_blank" rel="noopener" href="https://benlangmead.github.io/aws-indexes/k2">https://benlangmead.github.io/aws-indexes/k2</a></p>
<p>下载Standard 文件大小 90G</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">#自行下载命令</span><br><span class="line">wget -c https://genome-idx.s3.amazonaws.com/kraken/k2_standard_20240605.tar.gz</span><br></pre></td></tr></table></figure>

<p>#运行Kraken2：使用Kraken2对测序数据进行分类需要使用以下命令：</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">kraken2 --db &lt;path_to_database&gt; &lt;input_file&gt; --output &lt;output_file&gt;</span><br></pre></td></tr></table></figure>

<p>这里，<code>&lt;path_to_database&gt;</code>是参考数据库的路径，<code>&lt;input_file&gt;</code>是需要进行分类的输入文件，<code>&lt;output_file&gt;</code>是输出文件的名称。Kraken2将输出一个分类报告文件和一个序列文件。</p>
<p>需要注意的是kraken运行至少要提供数据库大小的内存大小（运行内存），因为它会把整个数据库载入内存后进行序列的注释，所以如果发现无法载入数据库的报错，可以尝试调大内存资源。</p>
<h2 id="单个比对"><a href="#单个比对" class="headerlink" title="单个比对"></a>单个比对</h2><p>第二行- -db 为存放参考索引的文件夹</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">kraken2 --threads 20 \</span><br><span class="line">    --db /public/home/dk_szy/songxudong/metagenomic/test \</span><br><span class="line">    --confidence 0.05 \</span><br><span class="line">    --output ./result/test.output \</span><br><span class="line">    --report ./report/test.kreport \</span><br><span class="line">    --paired \</span><br><span class="line">    ../rm_human/ERR1018185.1.fq \</span><br><span class="line">    ../rm_human/ERR1018185.2.fq</span><br></pre></td></tr></table></figure>

<h2 id="批量比对"><a href="#批量比对" class="headerlink" title="批量比对"></a>批量比对</h2><figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">mkdir kraken-result</span><br><span class="line"># 将工作目录设置为fastq文件所在的目录！！！！</span><br><span class="line">cd ./rm_human/</span><br><span class="line"></span><br><span class="line"># 将传入的参数赋值给变量！！！！！</span><br><span class="line">file1_pattern=&quot;.1.fq&quot;</span><br><span class="line">file2_pattern=&quot;.2.fq&quot;</span><br><span class="line"></span><br><span class="line"># 遍历所有以第一个参数模式结尾的文件</span><br><span class="line">for file1 in *$&#123;file1_pattern&#125;; do</span><br><span class="line">    # 从文件名中提取去掉模式后的部分</span><br><span class="line">    base=$(basename &quot;$file1&quot; $&#123;file1_pattern&#125;)</span><br><span class="line">    # 构建对应的第二个参数模式的文件名</span><br><span class="line">    file2=&quot;$&#123;base&#125;$&#123;file2_pattern&#125;&quot;</span><br><span class="line"></span><br><span class="line">    # 在后台执行检查和trim_galore命令</span><br><span class="line">    (</span><br><span class="line">        # 检查对应的文件是否存在</span><br><span class="line">        if [ -e &quot;$file2&quot; ]; then</span><br><span class="line">		echo &quot;找到名为 &quot;$base&quot; 的文件 $file1 对应 $file2 &quot;</span><br><span class="line">            # 如果存在，例如，执行trim_galore命令！！！！！！</span><br><span class="line">            #trim_galore -q 25 --phred33 --length 35 --stringency 3 --paired -o ../clean_data/ &quot;$file1&quot; &quot;$file2&quot;</span><br><span class="line">  		kraken2 --threads 20 \</span><br><span class="line">  	  --db /public/home/dk_szy/songxudong/metagenomic/test \</span><br><span class="line">  	  --confidence 0.05 \</span><br><span class="line">    	--output ../kraken-result/$&#123;base&#125;.output \</span><br><span class="line">    	--report ../kraken-result/$&#123;base&#125;.kreport \</span><br><span class="line">    	--paired \</span><br><span class="line">    	./$&#123;file1&#125; \</span><br><span class="line">    	./$&#123;file2&#125;</span><br><span class="line">		</span><br><span class="line">	echo &quot;文件 &quot;$base&quot; 运行完毕 &quot;</span><br><span class="line">        else</span><br><span class="line">            # 如果不存在，打印错误信息</span><br><span class="line">            echo &quot;错误: 未找到与 $file1 匹配的文件&quot;</span><br><span class="line">        fi</span><br><span class="line">    ) &amp;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 等待所有后台进程完成</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>

<h1 id="物种组成及丰度估计"><a href="#物种组成及丰度估计" class="headerlink" title="物种组成及丰度估计"></a>物种组成及丰度估计</h1><p>只能说宏基因组的教程也太水了，未找到对结果的处理部分</p>
<p>使用教程：<a target="_blank" rel="noopener" href="https://www.jianshu.com/nb/54122549">https://www.jianshu.com/nb/54122549</a></p>
<p>运行 bracken 进行各个分类水平物种丰度估计：</p>
<p>先安装 bracken</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda install bioconda::bracken</span><br></pre></td></tr></table></figure>

<p>运行 bracken 进行各个分类水平物种丰度估计：</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">mkdir out</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># 将工作目录设置为fastq文件所在的目录</span><br><span class="line">cd ./kraken-result/</span><br><span class="line"></span><br><span class="line"># 将传入的参数赋值给变量</span><br><span class="line">file1_pattern=&quot;.kreport&quot;</span><br><span class="line"></span><br><span class="line"># 遍历所有以第一个参数模式结尾的文件</span><br><span class="line">for file1 in *$&#123;file1_pattern&#125;; do</span><br><span class="line">    # 从文件名中提取去掉模式后的部分</span><br><span class="line">    base=$(basename &quot;$file1&quot; $&#123;file1_pattern&#125;)</span><br><span class="line">    </span><br><span class="line">    # 直接执行trim_galore命令，不需要检查对应的file2是否存在</span><br><span class="line">    (</span><br><span class="line">        echo &quot;找到名为 $base 的文件 $file1&quot;</span><br><span class="line">        #循环执行代码区</span><br><span class="line">        # 运行bracken</span><br><span class="line">bracken \</span><br><span class="line">-d /public/home/dk_szy/songxudong/metagenomic/test \</span><br><span class="line">-i $&#123;base&#125;.kreport \</span><br><span class="line">-o ../out/$&#123;base&#125;.bracken.S \</span><br><span class="line">-w ../out/$&#123;base&#125;.bracken.S.kreport \</span><br><span class="line">-l S \</span><br><span class="line">-t 20 </span><br><span class="line">    ) &amp;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 等待所有后台进程完成</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>

<h2 id="结果整理"><a href="#结果整理" class="headerlink" title="结果整理"></a>结果整理</h2><p>安装kraken-biom</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda install bioconda::kraken-biom</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line"># report文件合并成biom格式</span><br><span class="line">kraken-biom \</span><br><span class="line">./out/*.kreport \</span><br><span class="line">--max D \</span><br><span class="line">-o ./out/S.biom</span><br><span class="line"></span><br><span class="line"># biom转count表格</span><br><span class="line"># 注意：这里假设convert是指向biom转换为tsv的工具，不是ImageMagick的convert</span><br><span class="line"># 如果是ImageMagick的convert，那么下面的命令是错误的</span><br><span class="line"># 正确的命令应该是直接使用biom工具进行转换</span><br><span class="line">biom convert \</span><br><span class="line">-i ./out/S.biom \</span><br><span class="line">-o ./out/S.count.tsv.tmp \</span><br><span class="line">--to-tsv \</span><br><span class="line">--header-key taxonomy</span><br><span class="line"></span><br><span class="line"># 输出文件格式调整，补全物种名</span><br><span class="line">sed &#x27;s/; g__; s__/; g__; s__ /&#x27; ./out/S.count.tsv.tmp \</span><br><span class="line">&gt; ./out/S.taxID.count.tsv</span><br><span class="line"></span><br><span class="line"># taxonID 替换回拉丁名</span><br><span class="line">sed &#x27;/^#/! s/^[[0-9]]+\t\(.*[A-Za-z]__\([^\t;]\+\)\)$/\2\t\1/&#x27; \</span><br><span class="line">./out/S.taxID.count.tsv &gt; ./out/S.taxName.count.tsv</span><br><span class="line"></span><br><span class="line"># 保留丰度信息，用于后续绘图</span><br><span class="line">sed &#x27;1d; 2s/^#//&#x27; ./out/S.taxName.count.tsv | \</span><br><span class="line">awk -F $&#x27;\t&#x27; -v &#x27;OFS=\t&#x27; &#x27;&#123;$NF = &quot;&quot;; print $0&#125;&#x27; | \</span><br><span class="line">sed &#x27;s/\t$//&#x27; &gt; ./out/S.count.tsv</span><br></pre></td></tr></table></figure>

<p>输出结果：<br> S.taxName.count.tsv 为包含物种分类信息的丰度文件,可以用此结果绘制物种堆积图<br> S.count.tsv 为去除物种分类信息的丰度文件</p>
<p>暂未准备绘图代码<br><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/27313279-7f11db61e735cc99.png" alt="27313279-7f11db61e735cc99"></p>
<h1 id="a多样性"><a href="#a多样性" class="headerlink" title="a多样性"></a>a多样性</h1><p>教程中未具体讲解方法，但是提到   使用 vegan 及 phyloseq 两个 R 包进行多样性分析</p>
<p>使用16s做过的方法进行分析</p>
<p>OTU文件 S.count.tsv </p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240905155737677.png" alt="image-20240905155737677"></p>
<p>分组文件  group.txt</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240905155808792.png" alt="image-20240905155808792"></p>
<p>R语言</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">setwd<span class="punctuation">(</span><span class="string">&quot;D:/rtest/songtest/metagenomic&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 安装和加载必要的包</span></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>requireNamespace<span class="punctuation">(</span><span class="string">&quot;BiocManager&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">  install.packages<span class="punctuation">(</span><span class="string">&quot;BiocManager&quot;</span><span class="punctuation">)</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;phyloseq&quot;</span><span class="punctuation">)</span></span><br><span class="line">rm<span class="punctuation">(</span><span class="built_in">list</span><span class="operator">=</span>ls<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>vegan<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>reshape2<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>ggplot2<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>ggpubr<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>RColorBrewer<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">data <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span><span class="string">&quot;D:/rtest/songtest/metagenomic/S.count.txt&quot;</span><span class="punctuation">,</span>header<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span>row.names<span class="operator">=</span><span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">group <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span><span class="string">&quot;D:/rtest/songtest/metagenomic/group.txt&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#抽平</span></span><br><span class="line">otu <span class="operator">&lt;-</span> data</span><br><span class="line"><span class="comment">#求和查看每个样本的和</span></span><br><span class="line">colSums<span class="punctuation">(</span>data<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#使用该代码进行抽平</span></span><br><span class="line">otu_Flattening <span class="operator">=</span> as.data.frame<span class="punctuation">(</span>t<span class="punctuation">(</span>rrarefy<span class="punctuation">(</span>t<span class="punctuation">(</span>otu<span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">min</span><span class="punctuation">(</span>colSums<span class="punctuation">(</span>otu<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#查看抽平后的每个样本的和</span></span><br><span class="line">colSums<span class="punctuation">(</span>otu_Flattening<span class="punctuation">)</span></span><br><span class="line">data <span class="operator">&lt;-</span> otu_Flattening</span><br><span class="line"></span><br><span class="line">ttdata <span class="operator">&lt;-</span> t<span class="punctuation">(</span>data<span class="punctuation">)</span></span><br><span class="line">data<span class="operator">&lt;-</span>data<span class="operator">/</span>apply<span class="punctuation">(</span>data<span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="built_in">sum</span><span class="punctuation">)</span></span><br><span class="line">tdata<span class="operator">=</span>t<span class="punctuation">(</span>data<span class="punctuation">)</span></span><br><span class="line">a<span class="operator">&lt;-</span>as.data.frame<span class="punctuation">(</span>tdata<span class="punctuation">)</span></span><br><span class="line">a<span class="operator">=</span>as.data.frame<span class="punctuation">(</span>lapply<span class="punctuation">(</span>a<span class="punctuation">,</span><span class="built_in">as.numeric</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">shannon<span class="operator">&lt;-</span>diversity<span class="punctuation">(</span>a<span class="punctuation">,</span>index<span class="operator">=</span><span class="string">&quot;shannon&quot;</span><span class="punctuation">)</span></span><br><span class="line">simpson<span class="operator">&lt;-</span>diversity<span class="punctuation">(</span>a<span class="punctuation">,</span>index<span class="operator">=</span><span class="string">&quot;simpson&quot;</span><span class="punctuation">)</span></span><br><span class="line">Chao1  <span class="operator">&lt;-</span> estimateR<span class="punctuation">(</span>ttdata<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">ACE  <span class="operator">&lt;-</span> estimateR<span class="punctuation">(</span>ttdata<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">4</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line">invsimpson<span class="operator">&lt;-</span>diversity<span class="punctuation">(</span>a<span class="punctuation">,</span>index<span class="operator">=</span><span class="string">&quot;invsimpson&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">data_shannon<span class="operator">=</span>data.frame<span class="punctuation">(</span>shannon<span class="punctuation">)</span></span><br><span class="line">data_simpson<span class="operator">=</span>data.frame<span class="punctuation">(</span>simpson<span class="punctuation">)</span></span><br><span class="line">data_Chao1<span class="operator">=</span>data.frame<span class="punctuation">(</span>Chao1<span class="punctuation">)</span></span><br><span class="line">data_ACE<span class="operator">=</span>data.frame<span class="punctuation">(</span>ACE<span class="punctuation">)</span></span><br><span class="line">data_invsimpson<span class="operator">=</span>data.frame<span class="punctuation">(</span>invsimpson<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">spe_alpha<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span></span><br><span class="line">  data_shannon<span class="punctuation">,</span></span><br><span class="line">  data_simpson<span class="punctuation">,</span></span><br><span class="line">  data_invsimpson<span class="punctuation">,</span></span><br><span class="line">  data_Chao1<span class="punctuation">,</span></span><br><span class="line">  data_ACE<span class="punctuation">,</span></span><br><span class="line">  group</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h1 id="物种分类结果整理"><a href="#物种分类结果整理" class="headerlink" title="物种分类结果整理"></a>物种分类结果整理</h1><p>还是使用之前16s的方法</p>
<p>使用上面的 S.taxName.count.tsv 为包含物种分类信息的丰度文件,可以用此结果绘制物种堆积图</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240921154706720.png" alt="image-20240921154706720"></p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">setwd<span class="punctuation">(</span><span class="string">&quot;D:/rtest/songtest/metagenomic/lefse&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#taxon_data &lt;- read.csv(&quot;merged_otu_taxonomy.csv&quot;, stringsAsFactors = FALSE,row.names = 1)</span></span><br><span class="line">taxon_data <span class="operator">&lt;-</span> read.delim<span class="punctuation">(</span><span class="string">&quot;D:/rtest/songtest/metagenomic/lefse/S.taxName.count.tsv&quot;</span><span class="punctuation">,</span> row.names<span class="operator">=</span><span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span>kingdom <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span>phylum <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span><span class="built_in">class</span> <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span>order <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span>family <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span>genus <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">6</span><span class="punctuation">)</span></span><br><span class="line">taxon_data<span class="operator">$</span>species <span class="operator">&lt;-</span> sapply<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>taxon_data<span class="operator">$</span>taxonomy<span class="punctuation">,</span> <span class="string">&quot;;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> `[`<span class="punctuation">,</span> <span class="number">7</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">write.csv<span class="punctuation">(</span>taxon_data<span class="punctuation">,</span>file <span class="operator">=</span> <span class="string">&quot;otu-result.csv&quot;</span><span class="punctuation">)</span> <span class="comment">#保存</span></span><br></pre></td></tr></table></figure>

<p>导出分类好的  otu-result.csv  文件</p>
<h2 id="得到每个水平的结果"><a href="#得到每个水平的结果" class="headerlink" title="得到每个水平的结果"></a>得到每个水平的结果</h2><p>界（Kingdom）、门（Phylum）、纲（Class）、目（Order）、科（Family）、属（Genus）、种（Species）</p>
<p>16s只能注释到 属（Genus）水平的结果，虽然我们有属水平的文件，但是是没有相应的可用结果</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment"># 载入必要的库</span></span><br><span class="line">library<span class="punctuation">(</span>dplyr<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取数据</span></span><br><span class="line">otu_data <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span><span class="string">&quot;otu-result.csv&quot;</span><span class="punctuation">,</span> header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 移除Confidence列</span></span><br><span class="line">otu_data <span class="operator">&lt;-</span> otu_data<span class="punctuation">[</span><span class="operator">!</span><span class="built_in">names</span><span class="punctuation">(</span>otu_data<span class="punctuation">)</span> <span class="operator">%in%</span> <span class="string">&quot;Confidence&quot;</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 删除kingdom等于&quot;Unassigned&quot;的行</span></span><br><span class="line">otu_data <span class="operator">&lt;-</span> otu_data<span class="punctuation">[</span>otu_data<span class="operator">$</span>kingdom <span class="operator">!=</span> <span class="string">&quot;Unassigned&quot;</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 分类级别列表</span></span><br><span class="line">taxonomic_levels <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;phylum&quot;</span><span class="punctuation">,</span> <span class="string">&quot;class&quot;</span><span class="punctuation">,</span> <span class="string">&quot;order&quot;</span><span class="punctuation">,</span> <span class="string">&quot;family&quot;</span><span class="punctuation">,</span> <span class="string">&quot;genus&quot;</span><span class="punctuation">,</span> <span class="string">&quot;species&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 上一层级名称</span></span><br><span class="line">previous_level <span class="operator">&lt;-</span> <span class="string">&quot;kingdom&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 遍历每个分类级别，进行分组和求和，并保存到CSV文件</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="built_in">seq_along</span><span class="punctuation">(</span>taxonomic_levels<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  current_level <span class="operator">&lt;-</span> taxonomic_levels<span class="punctuation">[</span>i<span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># 确保当前层级和上一层级列存在并且数据类型正确</span></span><br><span class="line">  otu_data<span class="punctuation">[[</span>current_level<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="built_in">as.character</span><span class="punctuation">(</span>otu_data<span class="punctuation">[[</span>current_level<span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  otu_data<span class="punctuation">[[</span>previous_level<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="built_in">as.character</span><span class="punctuation">(</span>otu_data<span class="punctuation">[[</span>previous_level<span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># 对当前层级列进行分组并求和，同时保留上一层级信息</span></span><br><span class="line">  summary_data <span class="operator">&lt;-</span> otu_data <span class="operator">%&gt;%</span></span><br><span class="line">    group_by<span class="punctuation">(</span><span class="operator">!</span><span class="operator">!</span>sym<span class="punctuation">(</span>previous_level<span class="punctuation">)</span><span class="punctuation">,</span> <span class="operator">!</span><span class="operator">!</span>sym<span class="punctuation">(</span>current_level<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    summarise<span class="punctuation">(</span>across<span class="punctuation">(</span>where<span class="punctuation">(</span><span class="built_in">is.numeric</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="built_in">sum</span><span class="punctuation">,</span> na.rm <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    ungroup<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># 生成文件名</span></span><br><span class="line">  csv_filename <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>previous_level<span class="punctuation">,</span> <span class="string">&quot;_&quot;</span><span class="punctuation">,</span> current_level<span class="punctuation">,</span> <span class="string">&quot;.csv&quot;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># 保存结果到CSV文件</span></span><br><span class="line">  write.csv<span class="punctuation">(</span>summary_data<span class="punctuation">,</span> csv_filename<span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># 更新上一层级为当前层级</span></span><br><span class="line">  previous_level <span class="operator">&lt;-</span> current_level</span><br><span class="line"><span class="punctuation">&#125;</span></span><br></pre></td></tr></table></figure>

<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240812191527453.png" alt="image-20240812191527453"></p>
<h1 id="基因组组装"><a href="#基因组组装" class="headerlink" title="基因组组装"></a>基因组组装</h1><p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/77131fa96caa">https://www.jianshu.com/p/77131fa96caa</a></p>
<p><strong>使用 megahit 进行组装:</strong></p>
<p>安装</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">conda activate metagenomic</span><br><span class="line"></span><br><span class="line">conda install bioconda::megahit</span><br></pre></td></tr></table></figure>

<p>单个的示例</p>
<figure class="highlight csharp"><table><tr><td class="code"><pre><span class="line">megahit \</span><br><span class="line"><span class="number">-1</span> ./A1_1.fq.gz \ <span class="meta"># 输入，fq1</span></span><br><span class="line"><span class="number">-2</span> ./A1_2.fq.gz \ <span class="meta"># 输入，fq2</span></span><br><span class="line">--min-contig-len <span class="number">1000</span> \ <span class="meta"># contig最小长度</span></span><br><span class="line">--tmp-dir ./ \ <span class="meta"># 设置tmp目录</span></span><br><span class="line">--memory <span class="number">6</span> \ <span class="meta"># 内存占用</span></span><br><span class="line">--num-cpu-threads <span class="number">4</span> \ <span class="meta"># 线程数</span></span><br><span class="line">--<span class="keyword">out</span>-dir A1_megahit \ <span class="meta"># 输出目录</span></span><br><span class="line">--<span class="keyword">out</span>-prefix A1 <span class="meta"># 输出前缀</span></span><br><span class="line"><span class="meta">## 多组数据组装, 输入数据逗号分隔</span></span><br></pre></td></tr></table></figure>

<h2 id="批量组装"><a href="#批量组装" class="headerlink" title="批量组装"></a>批量组装</h2><p>要求：输出文件夹会自动创建，开始时不需要存在</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line"># 将工作目录设置为fastq文件所在的目录！！！！</span><br><span class="line">cd ./fastq-result/</span><br><span class="line"></span><br><span class="line"># 将传入的参数赋值给变量！！！！！</span><br><span class="line">file1_pattern=&quot;.1.fq&quot;</span><br><span class="line">file2_pattern=&quot;.2.fq&quot;</span><br><span class="line"></span><br><span class="line">#$&#123;file1&#125;</span><br><span class="line">#$&#123;file2&#125;</span><br><span class="line">#$&#123;base&#125;</span><br><span class="line"># 遍历所有以第一个参数模式结尾的文件</span><br><span class="line">for file1 in *$&#123;file1_pattern&#125;; do</span><br><span class="line">    # 从文件名中提取去掉模式后的部分</span><br><span class="line">    base=$(basename &quot;$file1&quot; $&#123;file1_pattern&#125;)</span><br><span class="line">    # 构建对应的第二个参数模式的文件名</span><br><span class="line">    file2=&quot;$&#123;base&#125;$&#123;file2_pattern&#125;&quot;</span><br><span class="line"></span><br><span class="line">    # 在后台执行检查和trim_galore命令</span><br><span class="line">    (</span><br><span class="line">        # 检查对应的文件是否存在</span><br><span class="line">        if [ -e &quot;$file2&quot; ]; then</span><br><span class="line">		echo &quot;找到名为 &quot;$base&quot; 的文件 $file1 对应 $file2 &quot;</span><br><span class="line">            # 如果存在，例如，执行trim_galore命令！！！！！！</span><br><span class="line">            #trim_galore -q 25 --phred33 --length 35 --stringency 3 --paired -o ../clean_data/ &quot;$file1&quot; &quot;$file2&quot;</span><br><span class="line">            megahit \</span><br><span class="line">-1 ./rm_human/$&#123;file1&#125; \</span><br><span class="line">-2 ./rm_human/$&#123;file2&#125; \</span><br><span class="line">--min-contig-len 1000 \</span><br><span class="line">--tmp-dir ./ \</span><br><span class="line">--memory 6 \</span><br><span class="line">--num-cpu-threads 20 \</span><br><span class="line">--out-dir ./megahit-result/  \</span><br><span class="line">--out-prefix $&#123;base&#125;</span><br><span class="line"></span><br><span class="line">        else</span><br><span class="line">            # 如果不存在，打印错误信息</span><br><span class="line">            echo &quot;错误: 未找到与 $file1 匹配的文件&quot;</span><br><span class="line">        fi</span><br><span class="line">    ) &amp;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"># 等待所有后台进程完成</span><br><span class="line">wait</span><br></pre></td></tr></table></figure>



<h1 id="Lefse"><a href="#Lefse" class="headerlink" title="Lefse"></a>Lefse</h1><p>参考： <a target="_blank" rel="noopener" href="https://blog.csdn.net/a852232394/article/details/139296579?spm=1001.2101.3001.6650.2&utm_medium=distribute.pc_relevant.none-task-blog-2~default~YuanLiJiHua~Position-2-139296579-blog-126683847.235%5Ev43%5Econtrol&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2~default~YuanLiJiHua~Position-2-139296579-blog-126683847.235%5Ev43%5Econtrol&utm_relevant_index=5">https://blog.csdn.net/a852232394/article/details/139296579?spm=1001.2101.3001.6650.2&amp;utm_medium=distribute.pc_relevant.none-task-blog-2%7Edefault%7EYuanLiJiHua%7EPosition-2-139296579-blog-126683847.235%5Ev43%5Econtrol&amp;depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7EYuanLiJiHua%7EPosition-2-139296579-blog-126683847.235%5Ev43%5Econtrol&amp;utm_relevant_index=5</a></p>
<p>需要3个输入文件</p>
<p>sample_table.csv<br><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240812220639621.png" alt="image-20240812220639621"></p>
<p>feature_table.csv</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240812220710558.png" alt="image-20240812220710558"></p>
<p>tax_table.csv</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240812220801693.png" alt="image-20240812220801693"></p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">rm(list=ls())</span><br><span class="line">pacman::p_load(tidyverse,microeco,magrittr)</span><br><span class="line"></span><br><span class="line">feature_table &lt;- read.csv(&#x27;feature_table.csv&#x27;, row.names = 1)</span><br><span class="line">sample_table &lt;- read.csv(&#x27;sample_table.csv&#x27;, row.names = 1)</span><br><span class="line">tax_table &lt;- read.csv(&#x27;tax_table.csv&#x27;, row.names = 1)</span><br><span class="line"></span><br><span class="line">head(feature_table)[,1:6]; head(sample_table); head(tax_table)[,1:6]</span><br><span class="line"></span><br><span class="line">dataset &lt;- microtable$new(sample_table = sample_table,</span><br><span class="line">                          otu_table = feature_table, </span><br><span class="line">                          tax_table = tax_table)</span><br><span class="line">dataset</span><br><span class="line"></span><br><span class="line">lefse &lt;- trans_diff$new(dataset = dataset, </span><br><span class="line">                        method = &quot;lefse&quot;, </span><br><span class="line">                        group = &quot;Group&quot;, </span><br><span class="line">                        #过少可增大下面选项</span><br><span class="line">                        alpha = 0.1, </span><br><span class="line">                        lefse_subgroup = NULL)</span><br><span class="line"></span><br><span class="line">write.csv(lefse$res_diff,file = &quot;lefse-lda.csv&quot;) #保存</span><br><span class="line">write.csv(lefse$abund_table,file = &quot;lefse-input.csv&quot;) #保存</span><br></pre></td></tr></table></figure>

<p>分析出初步结果</p>
<p>lefse-lda.csv  LDA，可自行画图</p>
<p>lefse-input.csv  lefse输入文件，可在在线网站进行可视化或者R可视化</p>
<h2 id="在线绘制"><a href="#在线绘制" class="headerlink" title="在线绘制"></a>在线绘制</h2><p>复制 lefse-input.csv 内容，进行修改</p>
<p><a target="_blank" rel="noopener" href="https://www.bic.ac.cn/BIC/#/">https://www.bic.ac.cn/BIC/#/</a></p>
<p>找到lefse选项</p>
<p>主要是更改第一行的列名，定义分组</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240813092603213.png" alt="image-20240813092603213"></p>
<p>检查数据可用后可直接出结果</p>
<p>（某些未知的菌可在原始数据中提前删除）</p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240813092855547.png" alt="image-20240813092855547"></p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240813092948111.png" alt="image-20240813092948111"></p>
<p>在线网站的编辑功能好像更强大了，作图方便建议采用</p>
<h2 id="R绘制"><a href="#R绘制" class="headerlink" title="R绘制"></a>R绘制</h2><p><a target="_blank" rel="noopener" href="https://blog.csdn.net/a852232394/article/details/139296579?spm=1001.2101.3001.6650.2&utm_medium=distribute.pc_relevant.none-task-blog-2~default~YuanLiJiHua~Position-2-139296579-blog-126683847.235%5Ev43%5Econtrol&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2~default~YuanLiJiHua~Position-2-139296579-blog-126683847.235%5Ev43%5Econtrol&utm_relevant_index=5">https://blog.csdn.net/a852232394/article/details/139296579?spm=1001.2101.3001.6650.2&amp;utm_medium=distribute.pc_relevant.none-task-blog-2~default~YuanLiJiHua~Position-2-139296579-blog-126683847.235%5Ev43%5Econtrol&amp;depth_1-utm_source=distribute.pc_relevant.none-task-blog-2~default~YuanLiJiHua~Position-2-139296579-blog-126683847.235%5Ev43%5Econtrol&amp;utm_relevant_index=5</a></p>
<p>使用  microeco包 自带的绘图，并不算好看，可进行美化或使用原始数据自行绘图</p>
<p>LDA</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">##use_number 为显示的个数</span><br><span class="line">##group_order 为自己的分组</span><br><span class="line">lefse$plot_diff_bar(use_number = 1:20, </span><br><span class="line">                    width = 0.8, </span><br><span class="line">                    group_order = c(&quot;subject-1&quot;, &quot;subject-2&quot;))</span><br></pre></td></tr></table></figure>

<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240813093433231.png" alt="image-20240813093433231"></p>
<p>lefse</p>
<figure class="highlight plaintext"><table><tr><td class="code"><pre><span class="line">library(ggtree)</span><br><span class="line">lefse$plot_diff_cladogram(use_taxa_num = 200, </span><br><span class="line">                       use_feature_num = 50, </span><br><span class="line">                       clade_label_level = 5, </span><br><span class="line">                       group_order =  c(&quot;subject-1&quot;, &quot;subject-2&quot;))</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20240813093404407.png" alt="image-20240813093404407"></p>
</article><div class="post-copyright"><div class="copyright-cc-box"><i class="anzhiyufont anzhiyu-icon-copyright"></i></div><div class="post-copyright__author_box"><a class="post-copyright__author_img" href="/" title="头像"><img class="post-copyright__author_img_back" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/img/touxiang.png" title="头像" alt="头像"><img class="post-copyright__author_img_front" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/img/touxiang.png" title="头像" alt="头像"></a><div class="post-copyright__author_name">Song Xudong</div><div class="post-copyright__author_desc">加油！</div></div><div class="post-copyright__post__info"><a class="post-copyright__original" title="该文章为原创文章，注意版权协议" href="http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/">原创</a><a class="post-copyright-title"><span onclick="rm.copyPageUrl('http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/')">宏基因组分析</span></a></div><div class="post-tools" id="post-tools"><div class="post-tools-left"><div class="rewardLeftButton"></div><div class="shareRight"><div class="share-link mobile"><div class="share-qrcode"><div class="share-button" title="使用手机访问这篇文章"><i class="anzhiyufont anzhiyu-icon-qrcode"></i></div><div class="share-main"><div class="share-main-all"><div id="qrcode" title="http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/"></div><div class="reward-dec">使用手机访问这篇文章</div></div></div></div></div><div class="share-link weibo"><a class="share-button" target="_blank" href="https://service.weibo.com/share/share.php?title=宏基因组分析&amp;url=http://example.com/2024/11/22/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/&amp;pic=/picture/fengmian/hong.jpg" rel="external nofollow noreferrer noopener"><i class="anzhiyufont anzhiyu-icon-weibo"></i></a></div><script>function copyCurrentPageUrl() {
  var currentPageUrl = window.location.href;
  var input = document.createElement("input");
  input.setAttribute("value", currentPageUrl);
  document.body.appendChild(input);
  input.select();
  input.setSelectionRange(0, 99999);
  document.execCommand("copy");
  document.body.removeChild(input);
}</script><div class="share-link copyurl"><div class="share-button" id="post-share-url" title="复制链接" onclick="copyCurrentPageUrl()"><i class="anzhiyufont anzhiyu-icon-link"></i></div></div></div></div></div><div class="post-copyright__notice"><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="http://example.com" target="_blank">宋旭东</a>！</span></div></div><div class="post-tools-right"><div class="tag_share"><div class="post-meta__box"><div class="post-meta__box__tag-list"><a class="post-meta__box__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/"><span class="tags-punctuation"> <i class="anzhiyufont anzhiyu-icon-tag"></i></span>生物信息学<span class="tagsPageCount">4</span></a><a class="post-meta__box__tags" href="/tags/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/"><span class="tags-punctuation"> <i class="anzhiyufont anzhiyu-icon-tag"></i></span>宏基因组<span class="tagsPageCount">1</span></a></div></div></div><div class="post_share"><div class="social-share" data-image="/picture/fengmian/MATLAB34224.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.cbd.int/butterfly-extsrc@1.1.3/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"/><script src="https://cdn.cbd.int/butterfly-extsrc@1.1.3/sharejs/dist/js/social-share.min.js" defer="defer"></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2024/11/18/Ribo-seq%E7%BF%BB%E8%AF%91%E7%BB%84%E5%AD%A6/"><img class="prev-cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/ribo-seq.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">翻译组分析流程</div></div></a></div><div class="next-post pull-right"><a href="/2025/01/14/%E5%9B%BD%E5%AE%B6%E5%A5%96%E5%AD%A6%E9%87%91/"><img class="next-cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/guojiang.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">国家奖学金</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="anzhiyufont anzhiyu-icon-thumbs-up fa-fw" style="font-size: 1.5rem; margin-right: 4px"></i><span>喜欢这篇文章的人也看了</span></div><div class="relatedPosts-list"><div><a href="/2024/11/16/AutoDock%E5%88%86%E5%AD%90%E5%AF%B9%E6%8E%A5/" title="AutoDock分子对接"><img class="cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/autodock.png" alt="cover"><div class="content is-center"><div class="date"><i class="anzhiyufont anzhiyu-icon-calendar-days fa-fw"></i> 2024-11-16</div><div class="title">AutoDock分子对接</div></div></a></div><div><a href="/2024/11/18/Ribo-seq%E7%BF%BB%E8%AF%91%E7%BB%84%E5%AD%A6/" title="翻译组分析流程"><img class="cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/ribo-seq.png" alt="cover"><div class="content is-center"><div class="date"><i class="anzhiyufont anzhiyu-icon-calendar-days fa-fw"></i> 2024-11-18</div><div class="title">翻译组分析流程</div></div></a></div><div><a href="/2024/11/16/%E8%BD%AC%E5%BD%95%E7%BB%84%E4%B8%8A%E6%B8%B8%E5%88%86%E6%9E%90/" title="转录组上游分析"><img class="cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/rna-seq.png" alt="cover"><div class="content is-center"><div class="date"><i class="anzhiyufont anzhiyu-icon-calendar-days fa-fw"></i> 2024-11-16</div><div class="title">转录组上游分析</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-content"><div class="author-info__sayhi" id="author-info__sayhi" onclick="anzhiyu.changeSayHelloText()"></div><div class="author-info-avatar"><img class="avatar-img" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/img/touxiang.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-status"><img class="g-status" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/08/24/64e6ce9c507bb.png" alt="status"/></div></div><div class="author-info__description"><div style="line-height:1.38;margin:0.6rem 0;text-align:justify;color:rgba(255, 255, 255, 0.8);">这有关于<b style="color:#fff">生物信息学</b>相关的学习经历和成果，还有<b style="color:#fff">个人经历</b>和<b style="color:#fff">日常分享</b>。</div><div style="line-height:1.38;margin:0.6rem 0;text-align:justify;color:rgba(255, 255, 255, 0.8);">相信你可以在这里找到对你有用的<b style="color:#fff">知识</b>和<b style="color:#fff">教程</b>。</div></div><div class="author-info__bottom-group"><a class="author-info__bottom-group-left" href="/"><h1 class="author-info__name">Song Xudong</h1><div class="author-info__desc">加油！</div></a><div class="card-info-social-icons is-center"><a class="social-icon faa-parent animated-hover" href="https://github.com/song-xudong" target="_blank" title="Github"><i class="anzhiyufont anzhiyu-icon-github"></i></a></div></div></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-bullhorn anzhiyu-shake"></i><span>公告</span></div><div class="announcement_content">博客还在建设中，更多内容正在迁移，欢迎你的来访！</div></div><div class="card-widget card-countdown"><div class="item-headline"><i></i><span></span></div><div class="item-content"><div class="cd-count-left">
  <span class="cd-text">距离</span>
  <span class="cd-name" id="eventName"></span>
  <span class="cd-time" id="daysUntil"></span>
  <span class="cd-date" id="eventDate"></span>
</div>
<div id="countRight" class="cd-count-right"></div>
</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-bars"></i><span>文章目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E4%B8%8B%E8%BD%BD%E6%95%B0%E6%8D%AE"><span class="toc-number">1.</span> <span class="toc-text">下载数据</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9F%A5%E7%9C%8B%E5%90%8E%E5%8F%B0%E4%BB%BB%E5%8A%A1"><span class="toc-number">1.0.1.</span> <span class="toc-text">查看后台任务</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#SRA%E6%96%87%E4%BB%B6%E8%BD%AC%E4%B8%BAFASTQ%E6%A0%BC%E5%BC%8F"><span class="toc-number">2.</span> <span class="toc-text">SRA文件转为FASTQ格式</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8D%95%E4%B8%AA%E8%BD%AC%E6%A0%BC%E5%BC%8F"><span class="toc-number">2.1.</span> <span class="toc-text">单个转格式</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%89%B9%E9%87%8F%E8%BD%AC%E6%A0%BC%E5%BC%8F"><span class="toc-number">2.2.</span> <span class="toc-text">批量转格式</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%B4%A8%E6%8E%A7"><span class="toc-number">3.</span> <span class="toc-text">质控</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#fastp%E8%B4%A8%E6%8E%A7"><span class="toc-number">3.1.</span> <span class="toc-text">fastp质控</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#multiqc%E6%B1%87%E6%80%BB%E8%B4%A8%E6%8E%A7%E7%BB%93%E6%9E%9C"><span class="toc-number">3.2.</span> <span class="toc-text">multiqc汇总质控结果</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8E%BB%E5%AE%BF%E4%B8%BB-%E6%96%B9%E6%B3%95%E4%B8%80"><span class="toc-number">4.</span> <span class="toc-text">去宿主-方法一</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%9E%84%E5%BB%BA%E7%B4%A2%E5%BC%95"><span class="toc-number">4.1.</span> <span class="toc-text">构建索引</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#bowtie2%E6%AF%94%E5%AF%B9-%E5%8D%95%E4%B8%AA"><span class="toc-number">4.2.</span> <span class="toc-text">bowtie2比对,单个</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#bowtie2%E6%AF%94%E5%AF%B9-%E6%89%B9%E9%87%8F"><span class="toc-number">4.3.</span> <span class="toc-text">bowtie2比对,批量</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8hisat2%E7%9B%B4%E6%8E%A5%E7%94%A8%E6%9E%84%E5%BB%BA%E5%A5%BD%E7%9A%84%E7%B4%A2%E5%BC%95%E8%BF%9B%E8%A1%8C%E6%AF%94%E5%AF%B9"><span class="toc-number">4.4.</span> <span class="toc-text">#使用hisat2直接用构建好的索引进行比对</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8E%BB%E5%AE%BF%E4%B8%BB-%E6%96%B9%E6%B3%95%E4%BA%8C-kneaddata"><span class="toc-number">5.</span> <span class="toc-text">#去宿主-方法二  kneaddata</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%89%A9%E7%A7%8D%E6%B3%A8%E9%87%8A%EF%BC%9Akraken2"><span class="toc-number">6.</span> <span class="toc-text">物种注释：kraken2</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8D%95%E4%B8%AA%E6%AF%94%E5%AF%B9"><span class="toc-number">6.1.</span> <span class="toc-text">单个比对</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%89%B9%E9%87%8F%E6%AF%94%E5%AF%B9"><span class="toc-number">6.2.</span> <span class="toc-text">批量比对</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%89%A9%E7%A7%8D%E7%BB%84%E6%88%90%E5%8F%8A%E4%B8%B0%E5%BA%A6%E4%BC%B0%E8%AE%A1"><span class="toc-number">7.</span> <span class="toc-text">物种组成及丰度估计</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E6%95%B4%E7%90%86"><span class="toc-number">7.1.</span> <span class="toc-text">结果整理</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#a%E5%A4%9A%E6%A0%B7%E6%80%A7"><span class="toc-number">8.</span> <span class="toc-text">a多样性</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%89%A9%E7%A7%8D%E5%88%86%E7%B1%BB%E7%BB%93%E6%9E%9C%E6%95%B4%E7%90%86"><span class="toc-number">9.</span> <span class="toc-text">物种分类结果整理</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BE%97%E5%88%B0%E6%AF%8F%E4%B8%AA%E6%B0%B4%E5%B9%B3%E7%9A%84%E7%BB%93%E6%9E%9C"><span class="toc-number">9.1.</span> <span class="toc-text">得到每个水平的结果</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%BB%84%E8%A3%85"><span class="toc-number">10.</span> <span class="toc-text">基因组组装</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%89%B9%E9%87%8F%E7%BB%84%E8%A3%85"><span class="toc-number">10.1.</span> <span class="toc-text">批量组装</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Lefse"><span class="toc-number">11.</span> <span class="toc-text">Lefse</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%9C%A8%E7%BA%BF%E7%BB%98%E5%88%B6"><span class="toc-number">11.1.</span> <span class="toc-text">在线绘制</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#R%E7%BB%98%E5%88%B6"><span class="toc-number">11.2.</span> <span class="toc-text">R绘制</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-history"></i><span>最近发布</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2025/03/16/MATLAB%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86(%E4%B8%89)/" title="MATLAB影像数据处理(三)"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/MATLAB34224.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="MATLAB影像数据处理(三)"/></a><div class="content"><a class="title" href="/2025/03/16/MATLAB%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86(%E4%B8%89)/" title="MATLAB影像数据处理(三)">MATLAB影像数据处理(三)</a><time datetime="2025-03-16T10:10:48.533Z" title="发表于 2025-03-16 18:10:48">2025-03-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/03/16/MATLAB%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86(%E4%BA%8C)/" title="MATLAB影像数据处理(二)"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/MATLAB34224.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="MATLAB影像数据处理(二)"/></a><div class="content"><a class="title" href="/2025/03/16/MATLAB%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86(%E4%BA%8C)/" title="MATLAB影像数据处理(二)">MATLAB影像数据处理(二)</a><time datetime="2025-03-16T10:04:03.969Z" title="发表于 2025-03-16 18:04:03">2025-03-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/02/24/MATLAB%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86(%E4%B8%80)/" title="MATLAB影像数据处理(一)"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/MATLAB34224.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="MATLAB影像数据处理(一)"/></a><div class="content"><a class="title" href="/2025/02/24/MATLAB%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86(%E4%B8%80)/" title="MATLAB影像数据处理(一)">MATLAB影像数据处理(一)</a><time datetime="2025-02-24T08:19:02.073Z" title="发表于 2025-02-24 16:19:02">2025-02-24</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/01/16/LLM%E5%A4%A7%E6%A8%A1%E5%9E%8B%EF%BC%88%E4%B8%80%EF%BC%89/" title="LLM大模型（一）"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/image-20250116135716298.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="LLM大模型（一）"/></a><div class="content"><a class="title" href="/2025/01/16/LLM%E5%A4%A7%E6%A8%A1%E5%9E%8B%EF%BC%88%E4%B8%80%EF%BC%89/" title="LLM大模型（一）">LLM大模型（一）</a><time datetime="2025-01-16T05:55:12.374Z" title="发表于 2025-01-16 13:55:12">2025-01-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/01/14/%E5%9B%BD%E5%AE%B6%E5%A5%96%E5%AD%A6%E9%87%91/" title="国家奖学金"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/picture/fengmian/guojiang.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="国家奖学金"/></a><div class="content"><a class="title" href="/2025/01/14/%E5%9B%BD%E5%AE%B6%E5%A5%96%E5%AD%A6%E9%87%91/" title="国家奖学金">国家奖学金</a><time datetime="2025-01-14T07:04:28.447Z" title="发表于 2025-01-14 15:04:28">2025-01-14</time></div></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2024 - 2025 By Song Xudong</div></div></footer></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="sidebar-site-data site-data is-center"><a href="/archives/" title="archive"><div class="headline">文章</div><div class="length-num">11</div></a><a href="/tags/" title="tag"><div class="headline">标签</div><div class="length-num">14</div></a><a href="/categories/" title="category"><div class="headline">分类</div><div class="length-num">3</div></a></div><span class="sidebar-menu-item-title">功能</span><div class="sidebar-menu-item"><a class="darkmode_switchbutton menu-child" href="javascript:void(0);" title="显示模式"><i class="anzhiyufont anzhiyu-icon-circle-half-stroke"></i><span>显示模式</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> Blog</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/archives/"><i class="anzhiyufont anzhiyu-icon-box-archive faa-tada" style="font-size: 0.9em;"></i><span> 总览</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/categories/"><i class="anzhiyufont anzhiyu-icon-shapes faa-tada" style="font-size: 0.9em;"></i><span> 分类</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/tags/"><i class="anzhiyufont anzhiyu-icon-tags faa-tada" style="font-size: 0.9em;"></i><span> 标签</span></a></li></ul></div><div class="menus_item"><a class="site-page faa-parent animated-hover" href="/about/"><span> 关于本人</span></a></div><div class="menus_item"><a class="site-page faa-parent animated-hover" href="/comments/"><span> 留言板</span></a></div></div><span class="sidebar-menu-item-title">标签</span><div class="card-tags"><div class="item-headline"></div><div class="card-tag-cloud"><a href="/tags/AI/" style="font-size: 0.88rem;">AI<sup>1</sup></a><a href="/tags/MATLAB/" style="font-size: 0.88rem;">MATLAB<sup>3</sup></a><a href="/tags/RNA-seq/" style="font-size: 0.88rem;">RNA-seq<sup>1</sup></a><a href="/tags/Ribo-seq/" style="font-size: 0.88rem;">Ribo-seq<sup>1</sup></a><a href="/tags/%E4%B8%80%E4%BD%9C/" style="font-size: 0.88rem;">一作<sup>1</sup></a><a href="/tags/%E4%BA%8C%E4%BD%9C/" style="font-size: 0.88rem;">二作<sup>1</sup></a><a href="/tags/%E5%88%86%E5%AD%90%E5%AF%B9%E6%8E%A5/" style="font-size: 0.88rem;">分子对接<sup>1</sup></a><a href="/tags/%E5%A4%A7%E5%AD%A6/" style="font-size: 0.88rem;">大学<sup>1</sup></a><a href="/tags/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/" style="font-size: 0.88rem;">宏基因组<sup>1</sup></a><a href="/tags/%E5%BD%B1%E5%83%8F%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86/" style="font-size: 0.88rem;">影像数据处理<sup>3</sup></a><a href="/tags/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0/" style="font-size: 0.88rem;">机器学习<sup>1</sup></a><a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" style="font-size: 0.88rem;">生物信息学<sup>4</sup></a><a href="/tags/%E8%8E%B7%E5%A5%96/" style="font-size: 0.88rem;">获奖<sup>1</sup></a><a href="/tags/%E8%AE%BA%E6%96%87/" style="font-size: 0.88rem;">论文<sup>2</sup></a></div></div><hr/></div></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="anzhiyufont anzhiyu-icon-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">繁</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="anzhiyufont anzhiyu-icon-circle-half-stroke"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="anzhiyufont anzhiyu-icon-arrows-left-right"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="anzhiyufont anzhiyu-icon-gear"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="anzhiyufont anzhiyu-icon-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div><div id="rightMenu"><div class="rightMenu-group rightMenu-small"><div class="rightMenu-item" id="menu-backward"><i class="anzhiyufont anzhiyu-icon-arrow-left"></i></div><div class="rightMenu-item" id="menu-forward"><i class="anzhiyufont anzhiyu-icon-arrow-right"></i></div><div class="rightMenu-item" id="menu-refresh"><i class="anzhiyufont anzhiyu-icon-arrow-rotate-right" style="font-size: 1rem;"></i></div><div class="rightMenu-item" id="menu-top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></div></div><div class="rightMenu-group rightMenu-line rightMenuPlugin"><div class="rightMenu-item" id="menu-copytext"><i class="anzhiyufont anzhiyu-icon-copy"></i><span>复制选中文本</span></div><div class="rightMenu-item" id="menu-pastetext"><i class="anzhiyufont anzhiyu-icon-paste"></i><span>粘贴文本</span></div><a class="rightMenu-item" id="menu-commenttext"><i class="anzhiyufont anzhiyu-icon-comment-medical"></i><span>引用到评论</span></a><div class="rightMenu-item" id="menu-newwindow"><i class="anzhiyufont anzhiyu-icon-window-restore"></i><span>新窗口打开</span></div><div class="rightMenu-item" id="menu-copylink"><i class="anzhiyufont anzhiyu-icon-link"></i><span>复制链接地址</span></div><div class="rightMenu-item" id="menu-copyimg"><i class="anzhiyufont anzhiyu-icon-images"></i><span>复制此图片</span></div><div class="rightMenu-item" id="menu-downloadimg"><i class="anzhiyufont anzhiyu-icon-download"></i><span>下载此图片</span></div><div class="rightMenu-item" id="menu-newwindowimg"><i class="anzhiyufont anzhiyu-icon-window-restore"></i><span>新窗口打开图片</span></div><div class="rightMenu-item" id="menu-search"><i class="anzhiyufont anzhiyu-icon-magnifying-glass"></i><span>站内搜索</span></div><div class="rightMenu-item" id="menu-searchBaidu"><i class="anzhiyufont anzhiyu-icon-magnifying-glass"></i><span>百度搜索</span></div><div class="rightMenu-item" id="menu-music-toggle"><i class="anzhiyufont anzhiyu-icon-play"></i><span>播放音乐</span></div><div class="rightMenu-item" id="menu-music-back"><i class="anzhiyufont anzhiyu-icon-backward"></i><span>切换到上一首</span></div><div class="rightMenu-item" id="menu-music-forward"><i class="anzhiyufont anzhiyu-icon-forward"></i><span>切换到下一首</span></div><div class="rightMenu-item" id="menu-music-playlist" onclick="window.open(&quot;https://y.qq.com/n/ryqq/playlist/8802438608&quot;, &quot;_blank&quot;);" style="display: none;"><i class="anzhiyufont anzhiyu-icon-radio"></i><span>查看所有歌曲</span></div><div class="rightMenu-item" id="menu-music-copyMusicName"><i class="anzhiyufont anzhiyu-icon-copy"></i><span>复制歌名</span></div></div><div class="rightMenu-group rightMenu-line rightMenuOther"><a class="rightMenu-item menu-link" id="menu-randomPost"><i class="anzhiyufont anzhiyu-icon-shuffle"></i><span>随便逛逛</span></a><a class="rightMenu-item menu-link" href="/categories/"><i class="anzhiyufont anzhiyu-icon-cube"></i><span>博客分类</span></a><a class="rightMenu-item menu-link" href="/tags/"><i class="anzhiyufont anzhiyu-icon-tags"></i><span>文章标签</span></a></div><div class="rightMenu-group rightMenu-line rightMenuOther"><a class="rightMenu-item" id="menu-copy" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-copy"></i><span>复制地址</span></a><a class="rightMenu-item" id="menu-commentBarrage" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-message"></i><span class="menu-commentBarrage-text">关闭热评</span></a><a class="rightMenu-item" id="menu-darkmode" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-circle-half-stroke"></i><span class="menu-darkmode-text">深色模式</span></a><a class="rightMenu-item" id="menu-translate" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-language"></i><span>轉為繁體</span></a></div></div><div id="rightmenu-mask"></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.cbd.int/@fancyapps/ui@5.0.28/dist/fancybox/fancybox.umd.js"></script><script src="https://cdn.cbd.int/instant.page@5.2.0/instantpage.js" type="module"></script><script src="https://cdn.cbd.int/vanilla-lazyload@17.8.5/dist/lazyload.iife.min.js"></script><script src="https://cdn.cbd.int/node-snackbar@0.1.16/dist/snackbar.min.js"></script><canvas id="universe"></canvas><script async src="https://npm.elemecdn.com/anzhiyu-theme-static@1.0.0/dark/dark.js"></script><script>// 消除控制台打印
var HoldLog = console.log;
console.log = function () {};
let now1 = new Date();
queueMicrotask(() => {
  const Log = function () {
    HoldLog.apply(console, arguments);
  }; //在恢复前输出日志
  const grt = new Date("10/01/2024 00:00:00"); //此处修改你的建站时间或者网站上线时间
  now1.setTime(now1.getTime() + 250);
  const days = (now1 - grt) / 1000 / 60 / 60 / 24;
  const dnum = Math.floor(days);
  const ascll = [
    `欢迎使用安知鱼!`,
    `生活明朗, 万物可爱`,
    `
        
       █████╗ ███╗   ██╗███████╗██╗  ██╗██╗██╗   ██╗██╗   ██╗
      ██╔══██╗████╗  ██║╚══███╔╝██║  ██║██║╚██╗ ██╔╝██║   ██║
      ███████║██╔██╗ ██║  ███╔╝ ███████║██║ ╚████╔╝ ██║   ██║
      ██╔══██║██║╚██╗██║ ███╔╝  ██╔══██║██║  ╚██╔╝  ██║   ██║
      ██║  ██║██║ ╚████║███████╗██║  ██║██║   ██║   ╚██████╔╝
      ╚═╝  ╚═╝╚═╝  ╚═══╝╚══════╝╚═╝  ╚═╝╚═╝   ╚═╝    ╚═════╝
        
        `,
    "已上线",
    dnum,
    "天",
    "©2024 By 安知鱼 V1.6.14",
  ];
  const ascll2 = [`NCC2-036`, `调用前置摄像头拍照成功，识别为【小笨蛋】.`, `Photo captured: `, `🤪`];

  setTimeout(
    Log.bind(
      console,
      `\n%c${ascll[0]} %c ${ascll[1]} %c ${ascll[2]} %c${ascll[3]}%c ${ascll[4]}%c ${ascll[5]}\n\n%c ${ascll[6]}\n`,
      "color:#425AEF",
      "",
      "color:#425AEF",
      "color:#425AEF",
      "",
      "color:#425AEF",
      ""
    )
  );
  setTimeout(
    Log.bind(
      console,
      `%c ${ascll2[0]} %c ${ascll2[1]} %c \n${ascll2[2]} %c\n${ascll2[3]}\n`,
      "color:white; background-color:#4fd953",
      "",
      "",
      'background:url("https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/tinggge.gif") no-repeat;font-size:450%'
    )
  );

  setTimeout(Log.bind(console, "%c WELCOME %c 你好，小笨蛋.", "color:white; background-color:#4f90d9", ""));

  setTimeout(
    console.warn.bind(
      console,
      "%c ⚡ Powered by 安知鱼 %c 你正在访问 Song Xudong 的博客.",
      "color:white; background-color:#f0ad4e",
      ""
    )
  );

  setTimeout(Log.bind(console, "%c W23-12 %c 你已打开控制台.", "color:white; background-color:#4f90d9", ""));

  setTimeout(
    console.warn.bind(console, "%c S013-782 %c 你现在正处于监控中.", "color:white; background-color:#d9534f", "")
  );
});</script><script async src="/anzhiyu/random.js"></script><div class="js-pjax"><input type="hidden" name="page-type" id="page-type" value="post"></div><script>var visitorMail = "songxudong@st.gxu.edu.cn";
</script><script async data-pjax src="https://cdn.cbd.int/anzhiyu-theme-static@1.0.0/waterfall/waterfall.js"></script><script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/qrcodejs/1.0.0/qrcode.min.js"></script><link rel="stylesheet" href="https://cdn.cbd.int/anzhiyu-theme-static@1.1.9/icon/ali_iconfont_css.css"><script src="/js/countdown.js"></script><link rel="stylesheet" href="https://cdn.cbd.int/anzhiyu-theme-static@1.0.0/aplayer/APlayer.min.css" media="print" onload="this.media='all'"><script src="https://cdn.cbd.int/anzhiyu-blog-static@1.0.1/js/APlayer.min.js"></script><script src="https://cdn.cbd.int/hexo-anzhiyu-music@1.0.1/assets/js/Meting2.min.js"></script><script src="https://cdn.cbd.int/pjax@0.2.8/pjax.min.js"></script><script>let pjaxSelectors = ["head > title","#config-diff","#body-wrap","#rightside-config-hide","#rightside-config-show",".js-pjax"]
var pjax = new Pjax({
  elements: 'a:not([target="_blank"])',
  selectors: pjaxSelectors,
  cacheBust: false,
  analytics: false,
  scrollRestoration: false
})

document.addEventListener('pjax:send', function () {
  // removeEventListener scroll 
  anzhiyu.removeGlobalFnEvent('pjax')
  anzhiyu.removeGlobalFnEvent('themeChange')

  document.getElementById('rightside').classList.remove('rightside-show')
  
  if (window.aplayers) {
    for (let i = 0; i < window.aplayers.length; i++) {
      if (!window.aplayers[i].options.fixed) {
        window.aplayers[i].destroy()
      }
    }
  }

  typeof typed === 'object' && typed.destroy()

  //reset readmode
  const $bodyClassList = document.body.classList
  $bodyClassList.contains('read-mode') && $bodyClassList.remove('read-mode')
})

document.addEventListener('pjax:complete', function () {
  window.refreshFn()

  document.querySelectorAll('script[data-pjax]').forEach(item => {
    const newScript = document.createElement('script')
    const content = item.text || item.textContent || item.innerHTML || ""
    Array.from(item.attributes).forEach(attr => newScript.setAttribute(attr.name, attr.value))
    newScript.appendChild(document.createTextNode(content))
    item.parentNode.replaceChild(newScript, item)
  })

  GLOBAL_CONFIG.islazyload && window.lazyLoadInstance.update()

  typeof panguInit === 'function' && panguInit()

  // google analytics
  typeof gtag === 'function' && gtag('config', '', {'page_path': window.location.pathname});

  // baidu analytics
  typeof _hmt === 'object' && _hmt.push(['_trackPageview',window.location.pathname]);

  typeof loadMeting === 'function' && document.getElementsByClassName('aplayer').length && loadMeting()

  // prismjs
  typeof Prism === 'object' && Prism.highlightAll()
})

document.addEventListener('pjax:error', e => {
  if (e.request.status === 404) {
    pjax.loadUrl('/404.html')
  }
})</script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><script charset="UTF-8" src="https://cdn.cbd.int/anzhiyu-theme-static@1.1.5/accesskey/accesskey.js"></script></div><div id="popup-window"><div class="popup-window-title">通知</div><div class="popup-window-divider"></div><div class="popup-window-content"><div class="popup-tip">你好呀</div><div class="popup-link"><i class="anzhiyufont anzhiyu-icon-arrow-circle-right"></i></div></div></div></body></html>